File: main_opencv.py
import sys
import os
import time
import cv2
import numpy as np


def has_display():
    return os.environ.get("OPENCV_ONNX_SHOW") == "1"


def result_path(file_name):
    output_dir = os.path.join(os.path.dirname(os.path.abspath(__file__)), "build")
    os.makedirs(output_dir, exist_ok=True)
    return os.path.join(output_dir, file_name)


def load_class_names(model_path):
    names_file_path = os.path.join(os.path.dirname(os.path.abspath(model_path)), "classes.names")
    class_names = {}
    if not os.path.exists(names_file_path):
        print(f"WARN: '{names_file_path}' not found. Using default numeric IDs.")
        return class_names

    with open(names_file_path, "r", encoding="utf-8") as f:
        for idx, line in enumerate(f):
            name = line.strip()
            if name:
                class_names[idx] = name
    return class_names


def hsv_to_rgb(h, s, v):
    rgb_max = v * 2.55
    rgb_min = rgb_max * (100 - s) / 100.0
    i = h // 60
    difs = h % 60
    rgb_adj = (rgb_max - rgb_min) * difs / 60.0

    if i == 0:
        r, g, b = rgb_max, rgb_min + rgb_adj, rgb_min
    elif i == 1:
        r, g, b = rgb_max - rgb_adj, rgb_max, rgb_min
    elif i == 2:
        r, g, b = rgb_min, rgb_max, rgb_min + rgb_adj
    elif i == 3:
        r, g, b = rgb_min, rgb_max - rgb_adj, rgb_max
    elif i == 4:
        r, g, b = rgb_min + rgb_adj, rgb_min, rgb_max
    else:
        r, g, b = rgb_max, rgb_min, rgb_max - rgb_adj

    return int(r), int(g), int(b)


def generate_palette_bgr(size_of_result):
    size = max(1, int(size_of_result))
    palette = []
    for i in range(size):
        h = int(360.0 / size * i)
        r, g, b = hsv_to_rgb(h, 100, 100)
        palette.append((b, g, r))
    return palette


def run_yolo11_seg_full_inference(model_path, image_path, show_boxes, conf_threshold, iou_threshold):
    class_names = load_class_names(model_path)
    num_classes = len(class_names) if class_names else 4
    palette_bgr = generate_palette_bgr(num_classes)

    image = cv2.imread(image_path)
    if image is None:
        print(f"ERROR: Could not read image '{image_path}'.")
        return
    original_image = image.copy()
    original_h, original_w = original_image.shape[:2]

    start_total_time = time.perf_counter()

    input_size = (640, 640)
    blob = cv2.dnn.blobFromImage(image, 1/255.0, input_size, swapRB=True, crop=False)

    net = cv2.dnn.readNetFromONNX(model_path)
    net.setInput(blob)
    out_names = net.getUnconnectedOutLayersNames()

    start_forward = time.perf_counter()
    outputs = net.forward(out_names)
    end_forward = time.perf_counter()
    forward_ms = (end_forward - start_forward) * 1000.0

    if outputs[0].shape[1] == 32:
        proto = outputs[0]
        preds = outputs[1]
    else:
        preds = outputs[0]
        proto = outputs[1]

    preds = np.squeeze(preds)
    if preds.shape[0] < preds.shape[1]:
        preds = preds.T

    proto = np.squeeze(proto)

    boxes = []
    confidences = []
    class_ids = []
    mask_coefficients = []

    x_factor = original_w / 640.0
    y_factor = original_h / 640.0

    for pred in preds:
        scores = pred[4 : 4 + num_classes]
        class_id = int(np.argmax(scores))
        confidence = float(scores[class_id])

        if confidence > conf_threshold:
            cx, cy, w, h = pred[0], pred[1], pred[2], pred[3]

            left = int((cx - w / 2) * x_factor)
            top = int((cy - h / 2) * y_factor)
            width = int(w * x_factor)
            height = int(h * y_factor)

            boxes.append([left, top, width, height])
            confidences.append(confidence)
            class_ids.append(class_id)
            mask_coefficients.append(pred[4 + num_classes :])

    indices = cv2.dnn.NMSBoxes(boxes, confidences, conf_threshold, iou_threshold)
    text_draw_queue = []

    if len(indices) > 0:
        indices = indices.flatten()
        c, proto_h, proto_w = proto.shape
        proto_flat = proto.reshape(c, -1)

        mask_overlay = np.zeros_like(original_image, dtype=np.uint8)

        for i in indices:
            left, top, width, height = boxes[i]
            class_id = class_ids[i]
            color = palette_bgr[class_id % len(palette_bgr)]

            coeffs = np.array(mask_coefficients[i])
            single_mask = np.dot(coeffs, proto_flat)
            single_mask = single_mask.reshape(proto_h, proto_w)

            # Sigmoid
            single_mask = 1 / (1 + np.exp(-single_mask))

            scale_mask = cv2.resize(single_mask, (640, 640), interpolation=cv2.INTER_LINEAR)

            # box 区域（无 padding，避免小目标异常）
            mx1 = max(0, int(left / x_factor))
            my1 = max(0, int(top / y_factor))
            mx2 = min(640, int((left + width) / x_factor))
            my2 = min(640, int((top + height) / y_factor))

            crop_mask = np.zeros_like(scale_mask)
            if mx2 > mx1 and my2 > my1:
                crop_mask[my1:my2, mx1:mx2] = scale_mask[my1:my2, mx1:mx2]

            actual_mask = cv2.resize(crop_mask, (original_w, original_h), interpolation=cv2.INTER_LINEAR)

            binary_mask = actual_mask > 0.5
            mask_overlay[binary_mask] = color

            if show_boxes:
                cv2.rectangle(mask_overlay, (left, top), (left + width, top + height), color, 2)

            label = f"{class_names.get(class_id, str(class_id))}: {confidences[i]:.2f}"
            font_scale = 0.55
            (text_w, text_h), baseline = cv2.getTextSize(label, cv2.FONT_HERSHEY_SIMPLEX, font_scale, 1)

            if show_boxes:
                text_x = left
                text_y = top - 8
                if text_y < text_h + 5:
                    text_y = top + height - 8
                    if text_y > original_h - 5:
                        text_y = top + text_h + 5
            else:
                text_x = left + (width // 2) - (text_w // 2)
                text_y = top + (height // 2) + (text_h // 2)

            text_x = max(5, min(text_x, original_w - text_w - 5))
            text_y = max(text_h + 5, min(text_y, original_h - baseline - 5))
            text_draw_queue.append((label, (text_x, text_y), font_scale))

        image = original_image.copy()
        cv2.addWeighted(mask_overlay, 0.4, image, 1.0, 0, image)

        for label, (x, y), scale in text_draw_queue:
            cv2.putText(image, label, (x, y), cv2.FONT_HERSHEY_SIMPLEX, scale, (0, 0, 0), 3, cv2.LINE_AA)
            cv2.putText(image, label, (x, y), cv2.FONT_HERSHEY_SIMPLEX, scale, (255, 255, 255), 1, cv2.LINE_AA)

        end_total_time = time.perf_counter()
        total_ms = (end_total_time - start_total_time) * 1000.0

        print("-------------------------------------------")
        print(f"PERF: Pure inference (forward) cost: {forward_ms:.2f} ms")
        print(f"PERF: Total algorithm pipeline cost: {total_ms:.2f} ms (including pre/post-processing)")
        print(f"INFO: Inference finished. Successfully detected {len(indices)} targets.")
        print("-------------------------------------------")
    else:
        print("WARN: No targets detected.")
        image = original_image.copy()

    output_name = result_path("result_opencv_dnn_python_cpu.png")
    cv2.imwrite(output_name, image)
    print(f"INFO: Result image saved to: {output_name}")
    if has_display():
        try:
            cv2.imshow("OpenCV 5 Python YOLO11-Seg", image)
            cv2.waitKey(0)
            cv2.destroyAllWindows()
        except cv2.error as exc:
            print(f"WARN: Could not open display window: {exc}")


if __name__ == "__main__":
    if len(sys.argv) < 6:
        print("ERROR: Insufficient arguments!")
        print("USAGE: python main_opencv.py <model_path.onnx> <image_path.png/.jpg> <True/False> <conf_threshold> <iou_threshold>")
        sys.exit(-1)

    model_arg = sys.argv[1]
    image_arg = sys.argv[2]
    show_boxes_arg = sys.argv[3].lower() in ['true', '1', 't', 'y', 'yes']
    conf_threshold_arg = float(sys.argv[4])
    iou_threshold_arg = float(sys.argv[5])

    print("INFO: Starting OpenCV 5 Python inference...")
    print(f" └─ Model: {model_arg}")
    print(f" └─ Image: {image_arg}")
    print(f" └─ Show Bounding Boxes: {show_boxes_arg}")
    print(f" └─ Confidence Threshold: {conf_threshold_arg}")
    print(f" └─ IoU NMS Threshold: {iou_threshold_arg}")

    run_yolo11_seg_full_inference(model_arg, image_arg, show_boxes_arg, conf_threshold_arg, iou_threshold_arg)

File: main_opencv.cpp
#include <opencv2/opencv.hpp>
#include <opencv2/dnn.hpp>
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <algorithm>
#include <cmath>
#include <chrono>
#include <cstdlib>
#include <limits.h>
#include <sys/stat.h>
#include <unistd.h>

struct TextObject {
    std::string label;
    cv::Point loc;
    double scale;
};

bool hasDisplay() {
    const char* showWindow = std::getenv("OPENCV_ONNX_SHOW");
    return showWindow != nullptr && std::string(showWindow) == "1";
}

void HSVtoRGB(int* r, int* g, int* b, int h, int s, int v) {
    float RGB_min, RGB_max;
    RGB_max = v * 2.55f;
    RGB_min = RGB_max * (100 - s) / 100.0f;

    int i = h / 60;
    int difs = h % 60;
    float RGB_Adj = (RGB_max - RGB_min) * difs / 60.0f;

    switch (i) {
    case 0:
        *r = static_cast<int>(RGB_max);
        *g = static_cast<int>(RGB_min + RGB_Adj);
        *b = static_cast<int>(RGB_min);
        break;
    case 1:
        *r = static_cast<int>(RGB_max - RGB_Adj);
        *g = static_cast<int>(RGB_max);
        *b = static_cast<int>(RGB_min);
        break;
    case 2:
        *r = static_cast<int>(RGB_min);
        *g = static_cast<int>(RGB_max);
        *b = static_cast<int>(RGB_min + RGB_Adj);
        break;
    case 3:
        *r = static_cast<int>(RGB_min);
        *g = static_cast<int>(RGB_max - RGB_Adj);
        *b = static_cast<int>(RGB_max);
        break;
    case 4:
        *r = static_cast<int>(RGB_min + RGB_Adj);
        *g = static_cast<int>(RGB_min);
        *b = static_cast<int>(RGB_max);
        break;
    default:
        *r = static_cast<int>(RGB_max);
        *g = static_cast<int>(RGB_min);
        *b = static_cast<int>(RGB_max - RGB_Adj);
        break;
    }
}

std::vector<cv::Scalar> paletteBgr(int sizeOfResult) {
    std::vector<cv::Scalar> colors;
    int size = std::max(1, sizeOfResult);
    colors.reserve(size);
    for (int i = 0; i < size; ++i) {
        int r = 0, g = 0, b = 0;
        int h = static_cast<int>(360.0 / size * i);
        HSVtoRGB(&r, &g, &b, h, 100, 100);
        colors.emplace_back(b, g, r);
    }
    return colors;
}

std::string executableDir() {
    char buffer[PATH_MAX] = {0};
    ssize_t length = readlink("/proc/self/exe", buffer, sizeof(buffer) - 1);
    std::string path = length > 0 ? std::string(buffer, static_cast<size_t>(length)) : std::string(".");
    size_t lastSlash = path.find_last_of("/\\");
    return lastSlash == std::string::npos ? "." : path.substr(0, lastSlash);
}

std::string resultPath(const std::string& fileName) {
    std::string outputDir = executableDir();
    mkdir(outputDir.c_str(), 0755);
    return outputDir + "/" + fileName;
}

// 根据 .onnx 的绝对路径，查找同级目录下的 classes.names
std::vector<std::string> loadClassNames(const std::string& modelPath) {
    std::vector<std::string> classNames;
    
    size_t lastSlash = modelPath.find_last_of("/\\");
    std::string baseDir = (lastSlash == std::string::npos) ? "" : modelPath.substr(0, lastSlash + 1);
    std::string namesFilePath = baseDir + "classes.names";

    std::ifstream ifs(namesFilePath);
    if (ifs.is_open()) {
        std::string line;
        while (std::getline(ifs, line)) {
            if (!line.empty() && line.back() == '\r') {
                line.pop_back();
            }
            if (!line.empty()) {
                classNames.push_back(line);
            }
        }
        ifs.close();
    } else {
        std::cout << "WARN: '" << namesFilePath << "' not found. Using default numeric IDs." << std::endl;
    }
    return classNames;
}

int main(int argc, char** argv) {
    // 🎯 1. 命令行参数校验
    if (argc < 6) {
        std::cerr << "ERROR: Insufficient arguments!" << std::endl;
        std::cerr << "USAGE: " << argv[0] << " <model_path.onnx> <image_path.png/.jpg> <true/false> <conf_threshold> <iou_threshold>" << std::endl;
        return -1;
    }

    std::string modelPath = argv[1];
    std::string imagePath = argv[2];
    std::string showBoxesStr = argv[3];
    
    std::transform(showBoxesStr.begin(), showBoxesStr.end(), showBoxesStr.begin(), ::tolower);
    bool showBoxes = (showBoxesStr == "true" || showBoxesStr == "1" || showBoxesStr == "yes");

    // 🎯 动态接收并解析外部传入的阈值参数
    float confThreshold = std::stof(argv[4]);
    float iouThreshold = std::stof(argv[5]); 

    // 动态提取类别名
    std::vector<std::string> classNames = loadClassNames(modelPath);
    int numClasses = static_cast<int>(classNames.size());

    std::cout << "INFO: Starting OpenCV 5 C++ inference..." << std::endl;
    std::cout << " └─ Model: " << modelPath << std::endl;
    std::cout << " └─ Image: " << imagePath << std::endl;
    std::cout << " └─ Show Bounding Boxes: " << (showBoxes ? "True" : "False") << std::endl;
    std::cout << " └─ Confidence Threshold: " << confThreshold << std::endl;
    std::cout << " └─ IoU NMS Threshold: " << iouThreshold << std::endl;

    // 2. 读取图像
    cv::Mat image = cv::imread(imagePath);
    if (image.empty()) {
        std::cerr << "ERROR: Could not read image: " << imagePath << std::endl;
        return -1;
    }
    int originalW = image.cols;
    int originalH = image.rows;

    // 启动全流程计时
    auto total_start = std::chrono::high_resolution_clock::now();

    cv::Mat maskOverlay = cv::Mat::zeros(image.size(), image.type());

    // 3. 图像预处理
    cv::Mat blob;
    cv::Size inputSize(640, 640);
    cv::dnn::blobFromImage(image, blob, 1.0 / 255.0, inputSize, cv::Scalar(), true, false);

    // 4. 加载模型并前向传播
    cv::dnn::Net net = cv::dnn::readNetFromONNX(modelPath);
    if (net.empty()) {
        std::cerr << "ERROR: Could not load ONNX model!" << std::endl;
        return -1;
    }
    net.setPreferableBackend(cv::dnn::DNN_BACKEND_OPENCV);
    net.setPreferableTarget(cv::dnn::DNN_TARGET_CPU);

    net.setInput(blob);
    std::vector<cv::Mat> outputs;

    // 单独测量前向传播（推理）耗时
    auto forward_start = std::chrono::high_resolution_clock::now();
    net.forward(outputs, net.getUnconnectedOutLayersNames());
    auto forward_end = std::chrono::high_resolution_clock::now();

    // 动态分流输出层
    cv::Mat preds, proto;
    if (outputs[0].size[1] == 32) {
        proto = outputs[0];
        preds = outputs[1];
    } else {
        preds = outputs[0];
        proto = outputs[1];
    }

    if (preds.dims == 3) {
        if (preds.size[1] < preds.size[2]) {
            preds = preds.reshape(1, preds.size[1]);
            cv::transpose(preds, preds);
        } else {
            preds = preds.reshape(1, preds.size[1]);
        }
    }

    std::vector<cv::Rect> boxes;
    std::vector<float> confidences;
    std::vector<int> classIds;
    std::vector<cv::Mat> maskCoefficients;

    float xFactor = originalW / 640.0f;
    float yFactor = originalH / 640.0f;

    // 5. 过滤目标
    for (int i = 0; i < preds.rows; ++i) {
        float* data = preds.ptr<float>(i);
        float* classesScores = data + 4;

        cv::Mat scoresMat(1, numClasses, CV_32FC1, classesScores);
        cv::Point classIdPoint;
        double maxClassScore;
        cv::minMaxLoc(scoresMat, 0, &maxClassScore, 0, &classIdPoint);

        if (maxClassScore > confThreshold) {
            float cx = data[0];
            float cy = data[1];
            float w = data[2];
            float h = data[3];

            int left = static_cast<int>((cx - w / 2.0f) * xFactor);
            int top = static_cast<int>((cy - h / 2.0f) * yFactor);
            int width = static_cast<int>(w * xFactor);
            int height = static_cast<int>(h * yFactor);

            boxes.push_back(cv::Rect(left, top, width, height));
            confidences.push_back(static_cast<float>(maxClassScore));
            classIds.push_back(classIdPoint.x);

            cv::Mat coeffs(1, 32, CV_32FC1, data + 4 + numClasses);
            maskCoefficients.push_back(coeffs.clone());
        }
    }

    std::vector<int> indices;
    // 使用外部参数传入的映射阈值
    cv::dnn::NMSBoxes(boxes, confidences, confThreshold, iouThreshold, indices);

    std::vector<cv::Scalar> colors = paletteBgr(numClasses);

    std::vector<TextObject> textQueue;

    // 6. 恢复掩码与计算绘制坐标
    if (!indices.empty()) {
        int protoChannels = proto.size[1];
        int protoH = proto.size[2];
        int protoW = proto.size[3];
        cv::Mat protoFlat = proto.reshape(1, protoChannels);

        for (int idx : indices) {
            cv::Rect box = boxes[idx];
            int classId = classIds[idx];
            cv::Scalar color = colors[classId % colors.size()];

            cv::Mat coeffs = maskCoefficients[idx];
            cv::Mat singleMaskMat = coeffs * protoFlat;
            singleMaskMat = singleMaskMat.reshape(1, protoH);

            for (int r = 0; r < protoH; ++r) {
                float* ptr = singleMaskMat.ptr<float>(r);
                for (int c = 0; c < protoW; ++c) {
                    ptr[c] = 1.0f / (1.0f + std::exp(-ptr[c]));
                }
            }

            cv::Mat scaleMask;
            cv::resize(singleMaskMat, scaleMask, cv::Size(640, 640));
            
            int mx1 = std::max(0, static_cast<int>(box.x / xFactor));
            int my1 = std::max(0, static_cast<int>(box.y / yFactor));
            int mx2 = std::min(640, static_cast<int>((box.x + box.width) / xFactor));
            int my2 = std::min(640, static_cast<int>((box.y + box.height) / yFactor));

            cv::Mat cropMask = cv::Mat::zeros(640, 640, CV_32FC1);
            if (mx2 > mx1 && my2 > my1) {
                scaleMask(cv::Rect(mx1, my1, mx2 - mx1, my2 - my1)).copyTo(cropMask(cv::Rect(mx1, my1, mx2 - mx1, my2 - my1)));
            }

            cv::Mat actualMask;
            cv::resize(cropMask, actualMask, image.size());
            cv::Mat binaryMask = actualMask > 0.5f;

            cv::Mat coloredMask(image.size(), image.type(), color);
            coloredMask.copyTo(maskOverlay, binaryMask);

            if (showBoxes) {
                cv::rectangle(maskOverlay, box, color, 2);
            }

            std::string nameLabel = (classId < classNames.size()) ? classNames[classId] : std::to_string(classId);
            std::string label = nameLabel + ": " + cv::format("%.2f", confidences[idx]);
            
            double fontScale = 0.55;
            int baseLine;
            cv::Size labelSize = cv::getTextSize(label, cv::FONT_HERSHEY_SIMPLEX, fontScale, 1, &baseLine);

            int textX = 0, textY = 0;
            if (showBoxes) {
                textX = box.x;
                textY = box.y - 8;
                if (textY < labelSize.height + 5) {
                    textY = box.y + box.height - 8;
                    if (textY > originalH - 5) {
                        textY = box.y + labelSize.height + 5;
                    }
                }
            } else {
                // 完全平台解耦的外接框定位
                textX = box.x + (box.width / 2) - (labelSize.width / 2);
                textY = box.y + (box.height / 2) + (labelSize.height / 2);
            }

            textX = std::max(5, std::min(textX, originalW - labelSize.width - 5));
            textY = std::max(labelSize.height + 5, std::min(textY, originalH - baseLine - 5));

            textQueue.push_back({label, cv::Point(textX, textY), fontScale});
        }

        // 7. 混合半透明掩码层
        float alpha = 0.4f;
        cv::addWeighted(maskOverlay, alpha, image, 1.0f, 0.0, image);

        // 8. 全透明悬浮黑边描边字体绘制
        for (const auto& textObj : textQueue) {
            cv::putText(image, textObj.label, textObj.loc, cv::FONT_HERSHEY_SIMPLEX, textObj.scale, cv::Scalar(0, 0, 0), 3, cv::LINE_AA);
            cv::putText(image, textObj.label, textObj.loc, cv::FONT_HERSHEY_SIMPLEX, textObj.scale, cv::Scalar(255, 255, 255), 1, cv::LINE_AA);
        }

        auto total_end   = std::chrono::high_resolution_clock::now();

        double forwardCost = std::chrono::duration<double, std::milli>(forward_end - forward_start).count();
        double totalCost = std::chrono::duration<double, std::milli>(total_end - total_start).count();

        std::cout << "-------------------------------------------" << std::endl;
        std::cout << "PERF: Pure inference (forward) cost: " << cv::format("%.2f", forwardCost) << " ms" << std::endl;
        std::cout << "PERF: Total algorithm pipeline cost: " << cv::format("%.2f", totalCost) << " ms (including pre/post-processing)" << std::endl;
        std::cout << "INFO: Inference finished. Successfully detected " << indices.size() << " targets." << std::endl;
        std::cout << "-------------------------------------------" << std::endl;
    } else {
        std::cout << "WARN: No targets detected." << std::endl;
    }

    std::string outName = resultPath("result_opencv_dnn_cpp_cpu.png");
    cv::imwrite(outName, image);
    std::cout << "INFO: Result image saved to: " << outName << std::endl;
    if (hasDisplay()) {
        try {
            cv::imshow("OpenCV 5 C++ YOLO11-Seg", image);
            cv::waitKey(0);
        } catch (const cv::Exception& e) {
            std::cout << "WARN: Could not open display window: " << e.what() << std::endl;
        }
    }

    return 0;
}
